### ISSUE TITLE
Rate limit chat

### ISSUE THREAD

**Issue by mithrand0** [ReactiveDrop/reactivedrop_public_src#518]
Some users have binds, that they spam.

I don't know if chat is already rate limited, but if not, it would be nice to have two options to setup this:

`rd_chat_ratelimit 12 limit the amount of messages per ratelimit period`.
`rd_chat_ratelimit_burst` 30 allow bursting over this limit before ratelimit is activated.
`rd_chat_ratelimit_period` 60 the window where to search for messages.

Is there already an option to do something like this?

**Comment by qq578023708** [comment]
An anti-flood plug-in in sourcemod implements this function.

**Comment by BenLubar** [comment]
Chat in the game only affects people who are in the same server. If someone is being abusive within a server, we already have tools for dealing with that.

**Comment by mithrand0** [comment]
> If someone is being abusive within a server, we already have tools for dealing with that.

I don't completely agree. Server owners are not moderators who are 24/7 present to deal with incidents. If a player is excessively spamming, it would be better to have them handled by a system, to prevent abuse, rather than a ban afterwards.

See it the same like hackers or cheaters, most of them get kicked and/or banned automatically when they do stuff on ranked, leaving only excessive cases to handle manually.

> An anti-flood plugin in sourcemod implements this function.

Regarding this, I don't like having a framework running in parallel for official servers. Not if the functionality can be easily implemented in the base game. With sourcemod, we leave it to the server owners to manage this. In the base game, all servers would have sane defaults to handle it, making an out of the box server a better experience.

**Comment by WhenTanksFly** [comment]
Vanilla already limits chat to about 1 message per second (hey can we have this not scale with adrenaline?); most servers with extra measures I've seen tighten that to a point where it's not possible to warn the team of several things happening in short succession. #worsethanvanilla
Also, some players won't give a damn to a single bind, even with caps lock; need entire chat filled with the same thing for them to have a chance to read it. Though I haven't tried repeating the same line 10 times in one message, maybe that'd work too...

**Comment by BenLubar** [comment]
Weird, it looks like it uses the logic for when an NPC can say a voice line to determine when a player can send a chat message.

https://github.com/ReactiveDrop/reactivedrop_public_src/blob/a8e01f52a9fb29aa72add3843f3ac1c8517bf92f/src/game/server/client.cpp#L151
